**Penerapan Algoritma _Clustering_ untuk Pengelompokkan Saham IDX Berdasarkan Indikator-indikator Fundamental - Submission Machine Learning Terapan Dicoding** 

oleh: Fikri Septrian Anggara (fikri_anggara_2c3r)

### Import library yang diperlukan

In [29]:
# untuk pengolahan data
import pandas as pd

# untuk visualisasi data
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme()

# untuk pembangunan klaster
from sklearn.cluster import KMeans
from sklearn.cluster import AgglomerativeClustering

# untuk visualisasi klaster
from scipy.cluster.hierarchy import dendrogram, linkage

# untuk scaling data
from sklearn.preprocessing import MinMaxScaler

# untuk menghitung jarak
from scipy.spatial.distance import pdist, cdist

# untuk menghitung evaluasi klaster
from sklearn.metrics import silhouette_score

# untuk melakukan reduksi dimensi
from sklearn.decomposition import PCA

### Menyiapkan Dataset
dataset diperoleh dari kaggle, terdapat dua dataset yang digunakan yaitu:
- [financial statement idx stocks](https://www.kaggle.com/datasets/greegtitan/financial-statement-idx-stocks?resource=download) (kaggle)
 - [daftar saham](https://www.kaggle.com/datasets/muamkh/ihsgstockdata?select=DaftarSaham.csv) (kaggle)

In [30]:
# load data
stockQuarter = pd.read_csv('data/quarter.csv')
masterStock = pd.read_csv('data/DaftarSaham.csv')

### Overview Data

In [31]:
# banyaknya baris dan kolom (baris, kolom)
print(stockQuarter.shape)
print(masterStock.shape)

(208691, 8)
(829, 14)


In [32]:
# ambil sampel random sebanyak 5 data
stockQuarter.sample(n=5)


symbol account                                          type  \
37067    BMSR      IS                                 Total Revenue   
35381    BLTZ      CF                      Preferred Stock Payments   
26794    BCAP      CF                        Cash Paymentsfor Loans   
166435   ROTI      CF                    Change In Dividend Payable   
16580    ASLC      CF  Other Cash Paymentsfrom Operating Activities   

          2021-09-30    2021-12-31    2022-03-31    2022-06-30  2022-09-30  
37067   7.150737e+11  1.072165e+12  1.326827e+12  1.138574e+12         NaN  
35381            NaN           NaN           NaN           NaN         NaN  
26794            NaN           NaN           NaN           NaN         NaN  
166435           NaN           NaN           NaN           NaN         NaN  
16580            NaN           NaN -1.181278e+10 -2.103036e+10         NaN

terdapat banyak **null value** pada data saham perkuarter

In [33]:
# ambil sampel random sebanyak 5 data
masterStock.sample(n=5)

Code                            Name ListingDate        Shares  \
751  TIRA              Tira Austenite Tbk  1993-07-27  5.880000e+08   
46   ARII            Atlas Resources Tbk.  2011-11-08  3.131000e+09   
240  DSNG     Dharma Satya Nusantara Tbk.  2013-06-14  1.059984e+10   
179  CBMF       Cahaya Bintang Medan Tbk.  2020-04-09  1.875000e+09   
302  GOLD  Visi Telekomunikasi Infrastruk  2010-07-07  1.277276e+09   

     ListingBoard                  Sector  LastPrice     MarketCap  \
751  Pengembangan             Industrials      408.0  2.399040e+11   
46   Pengembangan                  Energy      264.0  8.265840e+11   
240         Utama  Consumer Non-Cyclicals      640.0  6.783899e+12   
179  Pengembangan      Consumer Cyclicals       50.0  9.375000e+10   
302  Pengembangan         Infrastructures      494.0  6.309743e+11   

       MinutesFirstAdded   MinutesLastUpdated     HourlyFirstAdded  \
751  2021-11-01 09:00:00  2023-01-06 15:59:00  2020-04-16 09:00:00   
46   2021-11-01 09:00:00  2023-01-06 15:59:00  2020-04-16 09:00:00   
240  2021-11-01 09:00:00  2023-01-06 15:59:00  2020-04-16 09:00:00   
179  2021-11-01 09:00:00  2023-01-06 15:59:00  2020-04-16 09:00:00   
302  2021-11-01 09:00:00  2023-01-06 15:59:00  2020-04-16 09:00:00   

       HourlyLastUpdated DailyFirstAdded DailyLastUpdated  
751  2023-01-06 15:00:00      2001-04-16       2023-01-06  
46   2023-01-06 15:00:00      2011-11-08       2023-01-06  
240  2023-01-06 15:00:00      2013-06-14       2023-01-06  
179  2023-01-06 15:00:00      2020-04-09       2023-01-06  
302  2023-01-06 15:00:00      2010-07-07       2023-01-06

In [45]:
# 5 number summary stock quarter
stockQuarter.describe(include='all').T

count unique                top   freq                  mean  \
symbol       208691    544               AALI    391                   NaN   
account      208691      3                 BS  73461                   NaN   
type         208149    388  Interest Received   1080                   NaN   
2021-09-30  51722.0    NaN                NaN    NaN  1918202605901.322021   
2021-12-31  51538.0    NaN                NaN    NaN  2042450921973.788574   
2022-03-31  53449.0    NaN                NaN    NaN  1940242135220.246094   
2022-06-30  50375.0    NaN                NaN    NaN  2139261720062.068848   
2022-09-30  17013.0    NaN                NaN    NaN  3359261415683.166504   

                              std                min      25%            50%  \
symbol                        NaN                NaN      NaN            NaN   
account                       NaN                NaN      NaN            NaN   
type                          NaN                NaN      NaN            NaN   
2021-09-30  24798786684018.976562 -175313000000000.0      0.0   5837091260.0   
2021-12-31  26106014954333.765625 -438813037000000.0      0.0   7402772575.0   
2022-03-31  25585023607145.355469 -223695701000000.0      0.0   5526650874.0   
2022-06-30  26835599887835.515625 -194169000000000.0      0.0   7726820540.0   
2022-09-30  33720778198892.175781 -197134246000000.0  79886.0  20220521000.0   

                       75%                 max  
symbol                 NaN                 NaN  
account                NaN                 NaN  
type                   NaN                 NaN  
2021-09-30  233651664187.5  1637950171000000.0  
2021-12-31  257243562462.0  1725611128000000.0  
2022-03-31  214030861495.0  1734074740000000.0  
2022-06-30  265889690000.0  1785706841000000.0  
2022-09-30  557420000000.0  1839336498000000.0

- terdapat **544** buah saham emiten
- terdapat **388** variabel pada laporan keuangan
- terdapat **3** kategori akun yaitu balance sheet, cash-flow dan income statement

In [44]:
# 5 number summary master stock
masterStock.describe(include='all').T

count unique                      top freq  \
Code                  829    829                     AALI    1   
Name                  829    829  Astra Agro Lestari Tbk.    1   
ListingDate           829    713      2022-11-08 00:00:00    6   
Shares              829.0    NaN                      NaN  NaN   
ListingBoard          829      3             Pengembangan  422   
Sector                829     11       Consumer Cyclicals  142   
LastPrice           824.0    NaN                      NaN  NaN   
MarketCap           824.0    NaN                      NaN  NaN   
MinutesFirstAdded     787     60      2021-11-01 09:00:00  698   
MinutesLastUpdated    787     18      2023-01-06 15:59:00  759   
HourlyFirstAdded      806    106      2020-04-16 09:00:00  662   
HourlyLastUpdated     806     15      2023-01-06 15:00:00  759   
DailyFirstAdded       824    501               2001-04-16  225   
DailyLastUpdated      824      2               2023-01-06  822   

                                     mean                    std  \
Code                                  NaN                    NaN   
Name                                  NaN                    NaN   
ListingDate                           NaN                    NaN   
Shares                 10946944360.291918     45395931792.267258   
ListingBoard                          NaN                    NaN   
Sector                                NaN                    NaN   
LastPrice                     1363.453883            3733.432446   
MarketCap           10076631834955.722656  50531421356601.734375   
MinutesFirstAdded                     NaN                    NaN   
MinutesLastUpdated                    NaN                    NaN   
HourlyFirstAdded                      NaN                    NaN   
HourlyLastUpdated                     NaN                    NaN   
DailyFirstAdded                       NaN                    NaN   
DailyLastUpdated                      NaN                    NaN   

                             min             25%              50%  \
Code                         NaN             NaN              NaN   
Name                         NaN             NaN              NaN   
ListingDate                  NaN             NaN              NaN   
Shares                 3600000.0    1230839821.0     3128090000.0   
ListingBoard                 NaN             NaN              NaN   
Sector                       NaN             NaN              NaN   
LastPrice                   25.0          101.75            287.0   
MarketCap           9000000000.0  297528301025.0  1111304868160.0   
MinutesFirstAdded            NaN             NaN              NaN   
MinutesLastUpdated           NaN             NaN              NaN   
HourlyFirstAdded             NaN             NaN              NaN   
HourlyLastUpdated            NaN             NaN              NaN   
DailyFirstAdded              NaN             NaN              NaN   
DailyLastUpdated             NaN             NaN              NaN   

                                75%                 max  
Code                            NaN                 NaN  
Name                            NaN                 NaN  
ListingDate                     NaN                 NaN  
Shares                 9327465018.0     1184363929502.0  
ListingBoard                    NaN                 NaN  
Sector                          NaN                 NaN  
LastPrice                     975.0             38000.0  
MarketCap           4236952973400.0  1012951085850000.0  
MinutesFirstAdded               NaN                 NaN  
MinutesLastUpdated              NaN                 NaN  
HourlyFirstAdded                NaN                 NaN  
HourlyLastUpdated               NaN                 NaN  
DailyFirstAdded                 NaN                 NaN  
DailyLastUpdated                NaN                 NaN

- Terdapat 11 sektor pada master stok dengan **Consumer Cyclicals** adalah sektor yang paling banyak emitennya
- Terdapat 829 buah saham emiten

### Exploratory Data Analysis (EDA)
- pada data yang digunakan penulis hanya menggunakan data kuarter ke-3 2022 
variabel yang penulis gunakan untuk membangun klaster terdapat pada kolom `type` di data `stockQuarter`.
selain itu penulis juga ingin melihat pola pada tiap sektor, maka data `stockQuarter` perlu direshape

In [46]:
# variabel pada laporan keuangan
pd.DataFrame(pd.unique(stockQuarter['type']))

0
0                          Long Term Equity Investment
1                           Additional Paid In Capital
2                                       Long Term Debt
3                               Ordinary Shares Number
4                                         Total Assets
..                                                 ...
384                      Interest Income From Deposits
385                            Credit Losses Provision
386                              Other Interest Income
387  Interest Expense For Federal Funds Sold And Se...
388                     Total Money Market Investments

[389 rows x 1 columns]

In [54]:
# temp = stockQuarter[['account', 'type']].copy()
# temp = temp.drop_duplicates()
# temp.to_csv('data/financialStatement.csv')

terdapat 389 variabel pada laporan keuangan

In [47]:
# buat kolom baru, gabungan antara account dan type
stockQuarter['account_type'] = stockQuarter['account']+'_'+stockQuarter['type']
stockQuarter.head()

symbol account                         type    2021-09-30    2021-12-31  \
0   AALI      BS  Long Term Equity Investment  3.235200e+11  3.309040e+11   
1   AALI      BS   Additional Paid In Capital  3.878995e+12  3.878995e+12   
2   AALI      BS               Long Term Debt  5.709887e+12  2.131944e+12   
3   AALI      BS       Ordinary Shares Number  1.924688e+09  1.924688e+09   
4   AALI      BS                 Total Assets  2.969401e+13  3.039991e+13   

     2022-03-31    2022-06-30    2022-09-30                    account_type  
0  3.275800e+11  2.850690e+11  4.714630e+11  BS_Long Term Equity Investment  
1  3.878995e+12  3.878995e+12  3.878995e+12   BS_Additional Paid In Capital  
2  2.144732e+12  2.220370e+12  3.281008e+12               BS_Long Term Debt  
3  1.924688e+09  1.924688e+09  1.924688e+09       BS_Ordinary Shares Number  
4  3.123278e+13  3.023399e+13  3.263865e+13                 BS_Total Assets

### reshaping data

In [18]:
# ambil data 2021 saja
data2021 = stockQuarter[['symbol', 'account_type', '2022-06-30']].copy()
# reshaping data, kolom" data yang baru merupakan value dari 'type' dari df yang lama
dataReshaped = data2021.pivot_table(
    index='symbol',
    columns='account_type',
    values='2022-06-30'
).reset_index()
dataReshaped.head()

account_type symbol  BS_Accounts Payable  BS_Accounts Receivable  \
0              AALI         8.259340e+11            4.305280e+11   
1              ABBA         4.565925e+10            1.927453e+10   
2              ABDA         1.726138e+10            7.472096e+10   
3              ABMM         1.617792e+08            1.994225e+08   
4              ACES         1.728831e+11            7.668345e+10   

account_type  BS_Accumulated Depreciation  BS_Additional Paid In Capital  \
0                           -1.276215e+13                   3.878995e+12   
1                           -1.616807e+11                  -4.555371e+10   
2                           -1.051394e+11                   8.109426e+09   
3                           -7.378516e+08                   1.150872e+08   
4                           -1.918547e+12                   4.405749e+11   

account_type  BS_Allowance For Doubtful Accounts Receivable  \
0                                             -2.553900e+10   
1                                             -6.364641e+10   
2                                                       NaN   
3                                             -1.031222e+07   
4                                              5.312285e+06   

account_type  BS_Allowance For Loans And Lease Losses  \
0                                                 NaN   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   

account_type  BS_Assets Held For Sale  BS_Available For Sale Securities  \
0                                 NaN                               NaN   
1                                 NaN                      1.659867e+11   
2                                 NaN                     -2.732850e+09   
3                                 NaN                      2.346308e+07   
4                                 NaN                               NaN   

account_type  BS_Buildings And Improvements  ...  \
0                              4.876408e+12  ...   
1                              2.361747e+10  ...   
2                              9.103365e+10  ...   
3                              8.629613e+07  ...   
4                              2.333466e+12  ...   

account_type  IS_Selling General And Administration  \
0                                      1.818320e+11   
1                                      4.808846e+09   
2                                      2.573074e+10   
3                                      9.630067e+06   
4                                      8.751484e+10   

account_type  IS_Special Income Charges  IS_Tax Effect Of Unusual Items  \
0                          2.423000e+09                    7.975096e+08   
1                          0.000000e+00                    0.000000e+00   
2                          5.790000e+06                    2.316000e+06   
3                                   NaN                    0.000000e+00   
4                         -6.539966e+09                   -1.060284e+09   

account_type  IS_Tax Provision  IS_Tax Rate For Calcs  \
0                 1.628460e+11                    0.0   
1                 3.356401e+08                    0.0   
2                -7.352070e+08                    0.0   
3                 2.569592e+07                    0.0   
4                 1.740837e+10                    0.0   

account_type  IS_Total Premiums Earned  IS_Total Revenue  \
0                                  NaN      4.383457e+12   
1                                  NaN      3.270153e+10   
2                         1.570100e+11      1.933188e+11   
3                                  NaN      3.655084e+08   
4                                  NaN      1.681350e+12   

account_type  IS_Total Unusual Items  \
0                       2.423000e+09   
1                       0.000000e+00   
2                       5.790000e+06   
3 

### saham yang tersedia

In [17]:
pd.DataFrame(pd.unique(dataReshaped['symbol']))

0
0    AALI
1    ABBA
2    ABDA
3    ABMM
4    ACES
..    ...
498  WSKT
499  WTON
500  YPAS
501  ZBRA
502  ZINC

[503 rows x 1 columns]

In [19]:
# cek null value dari masing masing kolom
nulltable = pd.DataFrame(dataReshaped.isnull().sum().reset_index().iloc[1:])
nulltable.columns = ['financial statement', 'sum of null']
nulltable

financial statement  sum of null
1                              BS_Accounts Payable           17
2                           BS_Accounts Receivable           70
3                      BS_Accumulated Depreciation           10
4                    BS_Additional Paid In Capital            5
5    BS_Allowance For Doubtful Accounts Receivable          162
..                                             ...          ...
222                       IS_Total Premiums Earned          482
223                               IS_Total Revenue           14
224                         IS_Total Unusual Items          182
225      IS_Total Unusual Items Excluding Goodwill          182
226                                   IS_Write Off          398

[226 rows x 2 columns]

In [27]:
dataReshaped.isnull().sum()

account_type
symbol                                         0
BS_Accounts Payable                           17
BS_Accounts Receivable                        70
BS_Accumulated Depreciation                   10
BS_Additional Paid In Capital                  5
                                            ... 
IS_Total Premiums Earned                     482
IS_Total Revenue                              14
IS_Total Unusual Items                       182
IS_Total Unusual Items Excluding Goodwill    182
IS_Write Off                                 398
Length: 227, dtype: int64

In [20]:
usedColumns = [k for k in dataReshaped.columns.tolist() if 'BS_' in k][:15]


In [22]:
dataFinal = dataReshaped[usedColumns]
dataFinal

account_type  BS_Accounts Payable  BS_Accounts Receivable  \
0                    8.259340e+11            4.305280e+11   
1                    4.565925e+10            1.927453e+10   
2                    1.726138e+10            7.472096e+10   
3                    1.617792e+08            1.994225e+08   
4                    1.728831e+11            7.668345e+10   
..                            ...                     ...   
498                  7.007696e+12            1.501147e+13   
499                  2.239787e+12            1.119072e+12   
500                  1.161044e+10            8.208517e+10   
501                  5.400567e+11            6.963487e+11   
502                  1.101933e+11            5.717424e+10   

account_type  BS_Accumulated Depreciation  BS_Additional Paid In Capital  \
0                           -1.276215e+13                   3.878995e+12   
1                           -1.616807e+11                  -4.555371e+10   
2                           -1.051394e+11                   8.109426e+09   
3                           -7.378516e+08                   1.150872e+08   
4                           -1.918547e+12                   4.405749e+11   
..                                    ...                            ...   
498                         -4.096478e+12                   1.379202e+13   
499                         -1.791698e+12                   9.886334e+11   
500                         -1.646578e+11                   2.805402e+10   
501                         -3.866211e+11                   1.167524e+12   
502                         -4.867618e+11                   3.220000e+10   

account_type  BS_Allowance For Doubtful Accounts Receivable  \
0                                             -2.553900e+10   
1                                             -6.364641e+10   
2                                                       NaN   
3                                             -1.031222e+07   
4                                              5.312285e+06   
..                                                      ...   
498                                           -2.903400e+12   
499                                           -9.098793e+10   
500                                           -8.574100e+08   
501                                           -3.877312e+09   
502                                                     NaN   

account_type  BS_Allowance For Loans And Lease Losses  \
0                                                 NaN   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   
..                                                ...   
498                                               NaN   
499                                               NaN   
500                                               NaN   
501                                               NaN   
502                                               NaN   

account_type  BS_Assets Held For Sale  BS_Available For Sale Securities  \
0                                 NaN                               NaN   
1                                 NaN                      1.659867e+11   
2                                 NaN                     -2.732850e+09   
3                                 NaN                      2.346308e+07   
4                                 NaN                               NaN   
..                                ...                               ...   
498                               NaN                      8.829152e+11   
499                               NaN                               NaN   
500                               NaN                               NaN   
501                               NaN                               NaN   
502                               NaN                               NaN   

account_typ